src/raftstore/coprocessor/split_observer.rs

```diff
@@ -12,59 +12,49 @@
+use util::codec::table;
 
 use kvproto::raft_cmdpb::{SplitRequest, AdminRequest, Request, AdminResponse, Response,
                           AdminCmdType};
 use protobuf::RepeatedField;
-use util::codec::bytes;
 use std::result::Result as StdResult;
 
 /// `SplitObserver` adjusts the split key so that it won't seperate
-/// the data of a row into two region.
+/// the data of a row into two region. It adjusts the key according
+/// to the key format of `TiDB`.
 pub struct SplitObserver;
 
-pub const TABLE_PREFIX: &'static [u8] = b"t";
-pub const TABLE_ROW_MARK: &'static [u8] = b"_r";
+const VERSION_LENGTH: usize = 8;
 
 type Result<T> = StdResult<T, String>;
 
 impl SplitObserver {
-    fn extract_sql_key(&self, split: &SplitRequest) -> Result<Vec<u8>> {
+    fn on_split(&mut self, _: &mut ObserverContext, split: &mut SplitRequest) -> Result<()> {
         if !split.has_split_key() {
             return Err("split key is expected!".to_owned());
         }
-        let split_key = split.get_split_key();
-        bytes::decode_bytes(split_key)
-            .map_err(|e| format!("invalid key {:?}: {}", split_key, e))
-            .map(|r| r.0)
-    }
-
-    fn on_split(&mut self, _: &mut ObserverContext, split: &mut SplitRequest) -> Result<()> {
-        let key = try!(self.extract_sql_key(split));
-
-        // format of a key is TABLE_PREFIX + table_id + TABLE_ROW_MARK + handle or
-        // TABLE_PREFIX + table_id + TABLE_INDEX_MARK
+        let mut key = split.get_split_key().to_vec();
 
-        if !key.starts_with(TABLE_PREFIX) {
-            // if it's not a data key, just split.
+        if key.len() < VERSION_LENGTH {
+            // bypass non-mvcc key.
             return Ok(());
         }
-        // table id is a u64 and we only care about row data.
-        let mark_index = TABLE_PREFIX.len() + 8;
-        let new_key: &[u8];
-        if key.len() > mark_index && key[mark_index..].starts_with(TABLE_ROW_MARK) {
-            // handle is a u64 too.
-            let expected_len = TABLE_PREFIX.len() + 8 + TABLE_ROW_MARK.len() + 8;
-            if key.len() < expected_len {
-                return Err(format!("invalid key {:?}; len should be at least {}",
-                                   key,
-                                   expected_len));
-            }
-            new_key = &key[..expected_len];
+
+        // format of a key is TABLE_PREFIX + table_id + TABLE_ROW_MARK + handle + column_id
+        // + version or TABLE_PREFIX + table_id + TABLE_INDEX_MARK + index_id + values + version
+        // or meta_key + version
+
+        let table_prefix_len = table::TABLE_PREFIX.len() + table::ID_LEN;
+        if key.starts_with(table::TABLE_PREFIX) && key.len() > table::PREFIX_LEN + table::ID_LEN &&
+           key[table_prefix_len..].starts_with(table::RECORD_PREFIX_SEP) {
+            // row key, truncate to handle
+            key.truncate(table::PREFIX_LEN + table::ID_LEN);
```

Comment:
It's OK to truncate to a certain length for now, but in the future, we will use string handle, the prefix length will change.

We can append row key a column ID 0, so column key and row key would have the same length, then truncate suffix will ensure a row will not be split.